sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1012,85 +953,42 @@ DynRTDeviceBinaryImage::~DynRTDeviceBinaryImage() {
-    auto LockGuard = Ctx.getKernelProgramCache().acquireCachedPrograms();
-    auto It = NativePrograms.find(Prg);
-    if (It == NativePrograms.end()) {
-      if (DbgProgMgr > 0)
-        std::cerr << ">>> WARNING: flushSpecConstants requested on a PI "
-                     "program w/o known binary image\n";
-      return; // program origin is unknown
+    std::cerr << ">>> ProgramManager::flushSpecConstants(" << Prg.get()
+              << ",...)\n";
+  }
+  pi::PiProgram PrgHandle = Prg.getHandleRef();
+  // program_impl can't correspond to two different native programs
+  assert(!NativePrg || !PrgHandle || (NativePrg == PrgHandle));
+  NativePrg = NativePrg ? NativePrg : PrgHandle;
+
+  if (!Img) {
+    // caller hasn't provided the image object - find it
+    { // make sure NativePrograms map access is synchronized
+      ContextImplPtr Ctx = getSyclObjImpl(Prg.get_context());
+      auto LockGuard = Ctx->getKernelProgramCache().acquireCachedPrograms();
+      auto It = NativePrograms.find(NativePrg);
+      if (It == NativePrograms.end()) {
+        if (DbgProgMgr > 0)
+          std::cerr << ">>> WARNING: flushSpecConstants requested on a "
+                       "program w/o known binary image\n";
+        return; // program origin is unknown
+      }
+      Img = It->second;
     }
-    Img = It->second;
-  }
-  if (!Img->supportsSpecConstants())
-    // this device binary image does not support runtime setting of
-    // specialization constants; compiler must have generated default values
-    return;
-  OSModuleHandle ModuleHandle = Img->getOSModuleHandle();
-  if (DbgProgMgr > 2) {
-    Img->print();
-    std::cerr << "\n  OSModuleHandle:" << ModuleHandle << "\n";
-  }
-
-  auto It = SpecConstRegistry.find(ModuleHandle);
-  if (It == SpecConstRegistry.end()) {
-    if (DbgProgMgr > 0)
-      std::cerr << ">>> WARNING: flushSpecConstants no "
-                << "specialization contant registry for an OS module\n";
-    return;
-  }
-  const SpecConstMapTy &SCMap = It->second;
-  SpecConstNameIterator SCIt(*Img);
-
-  // iterate via all specialization constants the program's image depends on,
-  // and set each to current runtime value (if any)
-  for (const char *SCName = SCIt.next(); SCName; SCName = SCIt.next()) {
-    auto SCEntry = SCMap.find(SCName);
-    if (SCEntry == SCMap.end())
-      throw runtime_error("specialization constant " + std::string(SCName) +
-                              "not found in registry",
-                          PI_ERROR_UNKNOWN);
-    const spec_constant_impl &SC = SCEntry->second;
-    if (DbgProgMgr > 2) {
-      std::cerr << "  Spec const found: << " << SCName << " " << SC << "\n";
+    if (!Img->supportsSpecConstants()) {
+      if (DbgProgMgr > 0)
+        std::cerr << ">>> ProgramManager::flushSpecConstants: binary image "
+                  << &Img->getRawData() << " doesn't support spec constants\n";
+      // this device binary image does not support runtime setting of
```

Comment:
Do we throw exception somewhere if user sets spec constant for an image which doesn't support this feature?

Reply:
No. But exception can't be thrown here - the implemented spec has some gaps here.
What can happen in practice is AOT compilation, after which spec constants can't be set. On the other hand, we don't want to throw or change the program depending on JIT vs AOT scenario. Defaults will be used instead.

Reply:
added a comment

Reply:
So, there will be no diagnostic if a user uses AOT compiled binary but sets specialization constant. Is there any plan to address this problem?